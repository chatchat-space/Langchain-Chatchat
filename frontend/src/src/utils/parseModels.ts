import { CustomModels } from '@/types/settings';

export const parseModelString = (modelString: string = '') => {
  let models: CustomModels = [];
  let removeAll = false;
  const removedModels: string[] = [];
  const modelNames = modelString.split(/[,ï¼Œ]/).filter(Boolean);

  for (const item of modelNames) {
    const disable = item.startsWith('-');
    const nameConfig = item.startsWith('+') || item.startsWith('-') ? item.slice(1) : item;
    const [id, displayName] = nameConfig.split('=');

    if (disable) {
      // Disable all models.
      if (id === 'all') {
        removeAll = true;
      }
      removedModels.push(id);
      continue;
    }

    // Remove duplicate model entries.
    const existingIndex = models.findIndex(({ id: n }) => n === id);
    if (existingIndex !== -1) {
      models.splice(existingIndex, 1);
    }

    models.push({ displayName, id: id });
  }

  return {
    add: models,
    removeAll,
    removed: removedModels,
  };
};
