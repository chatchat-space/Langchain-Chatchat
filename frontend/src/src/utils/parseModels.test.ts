import { describe, expect, it } from 'vitest';

import { parseModelString } from './parseModels';

describe('parseModelString', () => {
  it('custom deletion, addition, and renaming of models', () => {
    const result = parseModelString(
      '-all,+llama,+claude-2，-gpt-3.5-turbo,gpt-4-1106-preview=gpt-4-turbo,gpt-4-1106-preview=gpt-4-32k',
    );

    expect(result).toMatchSnapshot();
  });

  it('duplicate naming model', () => {
    const result = parseModelString('gpt-4-1106-preview=gpt-4-turbo，gpt-4-1106-preview=gpt-4-32k');
    expect(result).toMatchSnapshot();
  });

  it('only add the model', () => {
    const result = parseModelString('model1,model2,model3，model4');

    expect(result).toMatchSnapshot();
  });
});
