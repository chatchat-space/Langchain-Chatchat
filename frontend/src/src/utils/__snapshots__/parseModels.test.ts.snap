// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseModelString > custom deletion, addition, and renaming of models 1`] = `
{
  "add": [
    {
      "displayName": undefined,
      "id": "llama",
    },
    {
      "displayName": undefined,
      "id": "claude-2",
    },
    {
      "displayName": "gpt-4-32k",
      "id": "gpt-4-1106-preview",
    },
  ],
  "removeAll": true,
  "removed": [
    "all",
    "gpt-3.5-turbo",
  ],
}
`;

exports[`parseModelString > duplicate naming model 1`] = `
{
  "add": [
    {
      "displayName": "gpt-4-32k",
      "id": "gpt-4-1106-preview",
    },
  ],
  "removeAll": false,
  "removed": [],
}
`;

exports[`parseModelString > only add the model 1`] = `
{
  "add": [
    {
      "displayName": undefined,
      "id": "model1",
    },
    {
      "displayName": undefined,
      "id": "model2",
    },
    {
      "displayName": undefined,
      "id": "model3",
    },
    {
      "displayName": undefined,
      "id": "model4",
    },
  ],
  "removeAll": false,
  "removed": [],
}
`;
